File: Chapter16/SampleViewAnimation/Java/SampleViewAnimation/app/src/main/java/com/ai/android/ExerciseSystemIntents/ViewAnimation3.java
package com.ai.android.ExerciseSystemIntents;

import android.graphics.Matrix;
import android.view.animation.Animation;
import android.view.animation.LinearInterpolator;
import android.view.animation.Transformation;

public class ViewAnimation3 extends Animation {

    private float centerX;
    private float centerY;

    public ViewAnimation3() {}

    @Override
    public void initialize(int width, int height, int parentWidth, int parentHeight) {
        super.initialize(width, height, parentWidth, parentHeight);
        centerX = width / 2.0f;
        centerY = height / 2.0f;
        setDuration(2500);
        setFillAfter(true);
        setInterpolator(new LinearInterpolator());
    }

    @Override
    protected void applyTransformation(float interpolatedTime, Transformation t) {
        final Matrix matrix = t.getMatrix();
        matrix.setScale(interpolatedTime, interpolatedTime);
        matrix.preTranslate(-centerX, -centerY);
        matrix.postTranslate(centerX, centerY);
    }
}

File: Chapter16/SampleViewAnimation/Kotlin/SampleViewAnimation/app/src/main/java/com/ai/android/sampleviewanimation/ViewAnimation3.kt
package com.ai.android.sampleviewanimation

import android.view.animation.Animation
import android.view.animation.LinearInterpolator
import android.view.animation.Transformation

class ViewAnimation3: Animation() {

    private var centerX: Float = 0.0f
    private var centerY: Float = 0.0f

    override fun initialize(width: Int, height: Int, parentWidth: Int, parentHeight: Int) {
        super.initialize(width, height, parentWidth, parentHeight)
        centerX = width / 2.0f
        centerY = height / 2.0f
        duration = 2500
        fillAfter = true
        interpolator = LinearInterpolator()
    }

    override fun applyTransformation(interpolatedTime: Float, t: Transformation?) {
        t?.apply {
            matrix?.apply {
                setScale(interpolatedTime, interpolatedTime)
                preTranslate(-centerX, -centerY)
                postTranslate(centerX, centerY)
            }
        }
    }
}